### PR TITLE
feat: More work and docs needed for jackson in native-image

### PR DESCRIPTION
**Pull request by johanandren**
Not entirely great but I think this improves on things quite a bit

### CODE REVIEW

File: native-image-tests/local-scala/src/main/scala/com/lightbend/JacksonCheck.scala [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        case "S" => South
+        case "W" => West
+      }
+    }
+  }
+
+  final case class Compass(@JsonProperty("currentDirection") currentDirection: Direction) extends JsonSerializable
+
+  // scala enums
+  // FIXME this one needs some very specific stuff for Scala enums, might not be worth it
```
Comment (johanandren) [inline native-image-tests/local-scala/src/main/scala/com/lightbend/JacksonCheck.scala:99]: This fails with reflective access to `$outer` so to support it we'd need to detect `Enumeration` subtype in `AkkaJacksonSerializationFeature` and handle specifically

File: native-image-tests/local-scala/src/main/scala/com/lightbend/JacksonCheck.scala [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+import com.fasterxml.jackson.databind.annotation.JsonSerialize
+import com.fasterxml.jackson.databind.deser.std.StdDeserializer
+import com.fasterxml.jackson.databind.ser.std.StdSerializer
+
+object JacksonCheck {
+
+  // from the akka-serialization-jackson docs
+
+  // type hierarchy / ADT
+  // FIXME: Still haven't figured out why @JsonCreator cannot be used here
```
Comment (johanandren) [inline native-image-tests/local-scala/src/main/scala/com/lightbend/JacksonCheck.scala:29]: This is really strange, I have verified that the constructor and parameter names are available for reflection, and for multi-param constructors it just works with nothing extra, but for some reason just `@JsonCreator` doesn't cut it here and a `@JsonProperty("fieldName")` is required.

I also ran with agent and tried with the generated reflect-config.json (which I expect to add the same or a bit less for the same class) and get the same error so it is something specifically about how Jackson reflectively finds single-param constructors with that annotation.

File: native-image-tests/local-scala/src/main/scala/com/lightbend/JacksonCheck.scala [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  final case class Zoo(@JsonProperty("primaryAttraction") primaryAttraction: Animal) extends JsonSerializable
+
+  // Note: native image needs additional marker trait on ADT supertype here or else
+  //       AkkaJacksonSerializationFeature has no way to detect and register the subtypes for
+  //       reflective access
```
Comment (johanandren) [inline native-image-tests/local-scala/src/main/scala/com/lightbend/JacksonCheck.scala:34]: This is really annoying, but can't see a way around except possibly doing classpath scanning, which we cannot do in a native image `Feature` so would have to be some other test infrastructure that users run to generate metadata files (a bit like the one I have in place internally in Akka).

File: akka-serialization-jackson/src/main/scala/akka/serialization/jackson/internal/AkkaJacksonSerializationFeature.scala [hunk 4]
```diff
@@ -28,28 +37,129 @@ final class AkkaJacksonSerializationFeature extends Feature {
...
+            log("Registering method " + clazz + "." + method.getName)
+            RuntimeReflection.register(method)
+            RuntimeReflection.registerAsQueried(method)
+          })
+
+      RuntimeReflection.registerAllConstructors(clazz)
+      RuntimeReflection.registerAllDeclaredConstructors(clazz)
+      // we still need explicit register of each constructor for some reason
+      util.Arrays.stream(clazz.getDeclaredConstructors).forEach { constructor =>
+        // FIXME this could probably be more selective
```
Comment (johanandren) [inline akka-serialization-jackson/src/main/scala/akka/serialization/jackson/internal/AkkaJacksonSerializationFeature.scala:108]: But unclear to me how Jackson chooses what constructor to use in the face of several, perhaps based on the incoming data and we really need to list all?

File: native-image-tests/local-scala/src/main/scala/com/lightbend/JacksonCheck.scala (hunk 3 above)
Comment (patriknw) [reply native-image-tests/local-scala/src/main/scala/com/lightbend/JacksonCheck.scala:44]: Feels like current state of art means that it is acceptable for users to have some trouble to solve themselves. Spring docs mention quite some caveats https://docs.spring.io/spring-boot/docs/current/reference/html/native-image.html#native-image.advanced.known-limitations

There I also noticed the link to https://github.com/oracle/graalvm-reachability-metadata
Is that something we can point at? I guess it's more related to libraries than user code, which is what you are trying to solve here.

Comment (johanandren) [reply native-image-tests/local-scala/src/main/scala/com/lightbend/JacksonCheck.scala:44]: I did figure out a way around it today, apparently you can register more type reachability callbacks from a reachability callback, so it now looks for interface constructor parameters and registers a new callback handler to find and register concrete types. Seems to work fine from the test coverage in local-scala native image test.

File: native-image-tests/local-scala/src/main/scala/com/lightbend/JacksonCheck.scala (hunk 1 above)
Comment (johanandren) [reply native-image-tests/local-scala/src/main/scala/com/lightbend/JacksonCheck.scala]: Decided to give up on Scala stdlib enums for now and documented that those doesn't work

File: native-image-tests/local-scala/src/main/scala/com/lightbend/JacksonCheck.scala (hunk 3 above)
Comment (johanandren) [reply native-image-tests/local-scala/src/main/scala/com/lightbend/JacksonCheck.scala:44]: For the reachability metadata I've gone with including what applies from the metadata repo, but coverage there is very spotty, so I think for now it is better, for user experience, with our own verified reflection metadata for transitive dependencies we say we support (unless they provide it themselves in their jars like Netty, which then just works tm).

File: native-image-tests/local-scala/src/main/scala/com/lightbend/JacksonCheck.scala (hunk 2 above)
Comment (patriknw) [reply native-image-tests/local-scala/src/main/scala/com/lightbend/JacksonCheck.scala]: hm, feels like the Jackson Scala module isn't used? Could it be that it's not included in native image and then falls back to regular Java Jackson, which requires the JsonCreator and has the param name problem?

Comment (johanandren) [reply native-image-tests/local-scala/src/main/scala/com/lightbend/JacksonCheck.scala]: Nope, that's not it, the objectMapper lists the same modules for normal JDK run as on native image.

File: akka-docs/src/main/paradox/additional/native-image.md [hunk 5]
```diff
@@ -38,10 +39,14 @@ do not need to provide metadata for them.
 ### Serialization with Jackson
 
 When using the built-in @apidoc[JsonSerializable] and @apidoc[CborSerializable] marker traits, message types are automatically added
-for reflective access by Akka. 
+for reflective access by Akka. But there are a few important caveats:
 
-If marker self-defined traits are being used instead, then the marker trait defined in `serialization-bindings`, as well as each 
-concrete message type (lookup and constructor), need to be added to the reflection metadata.
+Messages with only primitives, standard library your own or Akka provided types should work, but more complex
```
Comment (pvlugter) [inline akka-docs/src/main/paradox/additional/native-image.md:44]:
```suggestion
Messages with only primitives, standard library, your own or Akka provided types should work, but more complex
```

File: akka-docs/src/main/paradox/additional/native-image.md [hunk 6]
```diff
@@ -81,9 +86,5 @@ While Akka does not mandate a logger implementation, `logback-classic` is used i
 Because of this Akka provides reflection metadata for logback out of the box, however projects using it will need an extra
 native image flag `--initialize-at-build-time=ch.qos.logback`.
 
-FIXME maybe just provide the lazy appender out of the box as well? 
-      alternatively should we provide a lazy async logger appender or show it here?
-
-If using `ch.qos.logback.classic.AsyncAppender` you will need to declare your own lazy version of the appender (not starting any threads at native image build time).
-
-FIXME should we add link to working sample/config for logback?
+Special care needs to be taken if using the async logback appender, either avoid using it
```
Comment (pvlugter) [inline akka-docs/src/main/paradox/additional/native-image.md:89]:
```suggestion
Special care needs to be taken if using the async logback appender. Either avoid using
```

File: akka-serialization-jackson/src/main/scala/akka/serialization/jackson/internal/AkkaJacksonSerializationFeature.scala [hunk 7]
```diff
@@ -7,53 +7,225 @@ package akka.serialization.jackson.internal
...
+
+              if (parameterType.isInterface) {
+                // ADT or something like it, try to register concrete classes for that interface as well
+                access.registerSubtypeReachabilityHandler({ (access, clazz) =>
+                  registerTypeForJacksonSerialization(access, clazz)
+                }, parameterType)
+              }
+              if (classOf[scala.Enumeration#Value].isAssignableFrom(parameterType)) {
+                warning(
+                  "Saw a scala.Enumeration field this is not supported out of the box and will require adding manual metadata")
```
Comment (pvlugter) [inline akka-serialization-jackson/src/main/scala/akka/serialization/jackson/internal/AkkaJacksonSerializationFeature.scala:149]: Some form of punctuation here:

```suggestion
                  "Saw a scala.Enumeration field. This is not supported out of the box and will require adding manual metadata")
```

File: akka-serialization-jackson/src/main/scala/akka/serialization/jackson/internal/AkkaJacksonSerializationFeature.scala [hunk 8]
```diff
@@ -7,53 +7,225 @@ package akka.serialization.jackson.internal
...
+        }
+      }
+      // recursively register all enclosing classes/objects
+      registerEnclosing(clazz)
+    }
+  }
+
+  private def registerClassResourceForAccess(access: Feature.DuringAnalysisAccess, clazz: Class[_]): Unit = {
+    if (!alreadyRegisteredClassResources.contains(clazz.getName)) {
+      // Scala Jackson support needs the class files as resources as well
```
Comment (pvlugter) [inline akka-serialization-jackson/src/main/scala/akka/serialization/jackson/internal/AkkaJacksonSerializationFeature.scala:189]: 👍🏼

File: native-image-tests/local-scala/src/main/scala/com/lightbend/JacksonCheck.scala [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+  // these are auto-detected
+  final case class NestedType(someField: String)
+  final case class NestedType2(otherField: Double, yetAnother: Int)
+
+  // from the akka-serialization-jackson docs
+
+  // type hierarchy / ADT
+  // FIXME regular JDK doesn't need this @JsonCreator annotation but graal does
+  final case class Zoo(primaryAttraction: Animal) extends JsonSerializable
```
Comment (pvlugter) [inline native-image-tests/local-scala/src/main/scala/com/lightbend/JacksonCheck.scala:40]: FIXME still relevant? (no `@JsonCreator` annotation here)

File: native-image-tests/local-scala/src/main/scala/com/lightbend/JacksonCheck.scala [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize
+import com.fasterxml.jackson.databind.annotation.JsonSerialize
+import com.fasterxml.jackson.databind.deser.std.StdDeserializer
+import com.fasterxml.jackson.databind.ser.std.StdSerializer
+import com.fasterxml.jackson.module.scala.JsonScalaEnumeration
+
+object JacksonCheck {
+
+  // Note: class level JsonCreator does not work with native-image, it must be on the constructor
+  final case class SingleParam(field: String) extends JsonSerializable
```
Comment (pvlugter) [inline native-image-tests/local-scala/src/main/scala/com/lightbend/JacksonCheck.scala:28]: Is the note still relevant?

**Review by pvlugter — APPROVED** [review]
LGTM. That was more complicated, but great to have built-in support for this 👍🏼

**Comment by johanandren** [comment]
Green test run in https://github.com/akka/akka/actions/runs/8062783965

File: native-image-tests/local-scala/src/main/scala/com/lightbend/JacksonCheck.scala (hunk 2 above)
Comment (johanandren) [reply native-image-tests/local-scala/src/main/scala/com/lightbend/JacksonCheck.scala]: I explained elsewhere but it turns out that the Scala Jackson support reads/parses the actual classfiles as resources from the classpath, so needs a resource-config.json entry for graal. The `AkkaJacksonSerializationFeature` now automatically adds that for classes that should be serialisable with Jackson.

**Review by patriknw — APPROVED** [review]
LGTM

**Pull request merged**